src/coreclr/jit/importer.cpp

```diff
@@ -5716,7 +5716,30 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
     // Check legality only if an inline expansion is desirable.
     if (shouldExpandInline)
     {
-        if (isCastClass)
+        CORINFO_CLASS_HANDLE actualImplCls = NO_CLASS_HANDLE;
+        if (this->IsTargetAbi(CORINFO_NATIVEAOT_ABI) &&
+            ((helper == CORINFO_HELP_ISINSTANCEOFINTERFACE) || (helper == CORINFO_HELP_CHKCASTINTERFACE)) &&
+            (info.compCompHnd->getExactClasses(pResolvedToken->hClass, 1, &actualImplCls) == 1) &&
+            (actualImplCls != NO_CLASS_HANDLE) && impIsClassExact(actualImplCls))
```

Comment:
Is there any reason this optimization shouldn't kick in for any other potential EE that returns 1 from `getExactClasses`?

Reply:
@jakobbotsch purely throughput reasons - 2 ifs + non inlined call `getExactClasses` since I don't think it will be useful outside of NativeAOT anytime soon 🙂

Reply:
no strong opinion on that, will remove as part of some next PR